ext/app/electron/config.ts

```diff
@@ -0,0 +1,206 @@
+
+// Electron's app.getPath("userData") uses productName instead of name
+// but productName should be "full capitalized name", not ideal for naming our config directory.
+const DEFAULT_CONFIG_FILE = path.join(CONFIG_DIR, "config.ini")
+const NO_VALIDATION = () => true
+
+/**
+ * Suggest a value for an environment variable. If the variable is already set, do nothing. Otherwise set the value.
+ * @param name The name of the environment variable.
+ * @param value The value to suggest.
+ */
+function suggestEnv(name: string, value: string): void {
+  if (process.env[name] === undefined) {
+    process.env[name] = value;
+  }
+}
+
+type IConfig = {[key: string]: any}
+
+class Config implements IConfig {
+
+  [key: string]: any
+
+  constructor(c: IConfig) {
+    Object.assign(this, c)
+  }
+
+  /**
+   * Apply a configuration item by setting the corresponding environemt variable..
+   * The envvar, if already specified, has higher precedence over the config file.
+   *
+   * @param confKey The corresponding key in the config file.
+   * @param envKey The corresponding environment variable name.
+   * @param validator A predicate returning true if and only if the value to be used is valid. 
+   * @param defaultValue A default value to use when neither the environment variable nor the config file specifies a value, or when validation fails.
+   */
+  public apply(
+    confKey: string,
+    envKey: string,
+    validator: (value: string) => boolean,
+    defaultValue: string,
+  ): void {
+    let confValue: any = this
+    for (const segment of confKey.split(".")) {
+      confValue = confValue[segment]
+      if (confValue === undefined) {
+        break
+      }
+    }
+    let envValue = process.env[envKey]
+    if (envValue === undefined) {
+      if (confValue === undefined) {
+        log.info(`Neither ${confKey} nor $${envKey} is specified, using default value ${defaultValue}`)
+        process.env[envKey] = defaultValue
+        return
+      } else {
+        // envvar is undefined but config has it
+        if (!validator(confValue)) {
+          log.warn(`${confKey} has invalid value ${confValue}, using default value ${defaultValue}`)
+          process.env[envKey] = defaultValue
+        } else {
+          process.env[envKey] = confValue
+        }
+      }
+    } else {
+      // envvar is defined, ignore config
+      if (confValue !== undefined) {
+        log.warn(`${confKey} is overridden by $${envKey}`)
+      }
+      if (!validator(envValue)) {
+        log.warn(`$${envKey} has invalid value ${envValue}, using default value ${defaultValue}`)
+        process.env[envKey] = defaultValue
+      }
+    }
+  }
+}
+
+
+function loadConfigFile(filename: string = DEFAULT_CONFIG_FILE) {
+  let config: Config
+  try {
+    let configBuffer = fse.readFileSync(filename)
+    config = new Config(ini.parse(configBuffer.toString("utf8")))
+  } catch (err) {
+    log.warn(`Failed to read configuration file: ${err}`)
```

Comment:
Hmm if config file doesn't exist, seems ok to go ahead, but it for other errors, seems better to fail early?